authzed/api/v1/experimental_service.proto

```diff
@@ -136,3 +184,174 @@ message BulkExportRelationshipsResponse {
   Cursor after_result_cursor  = 1;
   repeated Relationship relationships = 2;
 }
+
+// Reflection types ////////////////////////////////////////////
+
+message ExperimentalReflectSchemaRequest {
+  Consistency consistency = 1;
+ 
+  // optional_filters defines optional filters that are applied in
+  // an OR fashion to the schema, before being returned
+  repeated ExpSchemaFilter optional_filters = 2;
+}
+
+message ExperimentalReflectSchemaResponse {
+  // definitions are the definitions defined in the schema.
+  repeated ExpDefinition definitions = 1;
+
+  // caveats are the caveats defined in the schema.
+  repeated ExpCaveat caveats = 2;
+
+  // read_at is the ZedToken at which the schema was read.
+  ZedToken read_at = 3;
+}
+
+message ExpSchemaFilter {
+  enum KindFilter {
+    KIND_FILTER_UNSPECIFIED = 0;
+    KIND_FILTER_DEFINITION = 1;
+    KIND_FILTER_CAVEAT = 2;
+    KIND_FILTER_RELATION = 3;
+    KIND_FILTER_PERMISSION = 4;
+  }
+
+  // optional_definition_name_match is a regex that is matched against the definition or caveat name.
+  // If not specified, will be ignored.
+  string optional_definition_name_match = 1;
+
+  // optional_relation_or_permission_name_match is a regex that is matched against the relation or permission name.
+  // If not specified, will be ignored.
+  string optional_relation_or_permission_name_match = 2;
+
+  // kind_filters is a list of kinds to filter on. If not specified, will be ignored. If multiple are specified,
+  // the filter will be applied in an OR fashion.
+  repeated KindFilter kind_filters = 3;
```

Comment:
@josephschorr when `kind==caveat`, the `optional_definition` makes sense, but `optional_relation` does not. IMHO a more type-safe way to model this is to have multiple types of `KindFilter` (`CaveatFilter`, `ObjectFilter`, `RelationFilter`), then remove both optional arguments, and only have `repeated KindFilter kind_filters` be a one of those types. The part I'm not sure is if you can create a repeated oneof

However, the `ExperimentalReflectSchemaRequest` already has a repeated `ExpSchemaFilter`, so why having the repeated `KindFilter` in `ExpSchemaFilter`? If superfluous, we could make `kind_filters` as I described and drop the repeated.

Reply:
> The part I'm not sure is if you can create a repeated oneof

You cannot; we'd have to create another message for the field and then repeat that

Reply:
What are you thoughts on this I wrote?

>However, the ExperimentalReflectSchemaRequest already has a repeated ExpSchemaFilter, so why having the repeated KindFilter in ExpSchemaFilter? If superfluous, we could make kind_filters as I described and drop the repeated.

Reply:
We can likely drop the repeated from in here, yes